File: 20-validParentheses/Cpp/validParentheses.cpp
#include <string>
#include <vector>
#include <iostream>

class Solution
{
public:
    bool isValid(std::string s)
    {
        std::vector<char> seen;

        for (int i = 0; i < s.size(); i++)
        {
            switch (s.at(i))
            {
            case '[':
                seen.push_back('[');
                break;
            case '{':
                seen.push_back('{');
                break;
            case '(':
                seen.push_back('(');
                break;
            case ']':
                if (seen.size() == 0 || seen[seen.size() - 1] != '[')
                {
                    return false;
                }
                seen.pop_back();
                break;
            case ')':
                if (seen.size() == 0 || seen[seen.size() - 1] != '(')
                {
                    return false;
                }
                seen.pop_back();
                break;
            case '}':
                if (seen.size() == 0 || seen[seen.size() - 1] != '{')
                {
                    return false;
                }
                seen.pop_back();
                break;
            }
        }
        return seen.size() == 0;
    }
};

File: 20-validParentheses/Go/validParentheses.go
package validParentheses

func isValid(s string) bool {
	seen := []rune{}

	for _, bracket := range s {

		switch bracket {
		case '[':
			seen = append(seen, '[')
		case '{':
			seen = append(seen, '{')
		case '(':
			seen = append(seen, '(')
		case ']':
			if len(seen) == 0 || seen[len(seen)-1] != '[' {
				return false
			}
			seen = seen[:len(seen)-1]
		case '}':
			if len(seen) == 0 || seen[len(seen)-1] != '{' {
				return false
			}
			seen = seen[:len(seen)-1]
		case ')':

			if len(seen) == 0 || seen[len(seen)-1] != '(' {
				return false
			}
			seen = seen[:len(seen)-1]
		}
	}
	return len(seen) == 0
}